### PR TITLE
Use pki-types pem decoding

### PR DESCRIPTION
**Pull request by ctz**
~~I was hoping to rely on cargo-semver-checks here to check for API regressions, alas https://github.com/obi1kenobi/cargo-semver-checks/issues/609 means this is not possible.~~

### CODE REVIEW

File: src/pemfile.rs [hunk 1]
```diff
@@ -1,59 +1,17 @@
...
-    ///
-    /// Appears as "X509 CRL" in PEM files.
-    Crl(CertificateRevocationListDer<'static>),
-
-    /// A Certificate Signing Request; as specified in RFC 2986
-    ///
-    /// Appears as "CERTIFICATE REQUEST" in PEM files.
-    Csr(CertificateSigningRequestDer<'static>),
-}
+use pki_types::pem as parent;
```
Comment (djc) [inline src/pemfile.rs:9]: Don't love the `parent` name here.

File: src/lib.rs [hunk 2]
```diff
@@ -104,10 +104,7 @@ pub fn private_key(rd: &mut dyn io::BufRead) -> Result<Option<PrivateKeyDer<'sta
             Item::Pkcs1Key(key) => return Ok(Some(key.into())),
             Item::Pkcs8Key(key) => return Ok(Some(key.into())),
             Item::Sec1Key(key) => return Ok(Some(key.into())),
-            Item::X509Certificate(_)
-            | Item::SubjectPublicKeyInfo(_)
-            | Item::Crl(_)
-            | Item::Csr(_) => continue,
+            _ => continue,
```
Comment (djc) [inline src/lib.rs:107]: Perhaps extract these fixes in a separate commit, since this seems like a logically independent change?

File: src/lib.rs [hunk 3]
```diff
@@ -72,6 +67,11 @@ use pki_types::{
     PrivatePkcs8KeyDer, PrivateSec1KeyDer, SubjectPublicKeyInfoDer,
 };
 
+mod pemfile;
+#[cfg(feature = "std")]
+pub use pemfile::{read_all, read_one};
+pub use pemfile::{read_one_from_slice, Error, Item};
```
Comment (djc) [inline src/lib.rs:73]: Suggest moving this code in a separate commit as well, to make obvious that it just moves this down by 8 lines.

File: src/pemfile.rs [hunk 4]
```diff
@@ -72,29 +30,57 @@ pub enum Error {
...
-        } else {
-            None
-        };
-
-        match read_one_impl(next_line, &mut section, &mut b64buf)? {
-            ControlFlow::Continue(()) => continue,
-            ControlFlow::Break(item) => return Ok(item.map(|item| (item, input))),
-        }
+pub fn read_one_from_slice(input: &[u8]) -> Result<Option<(Item, &[u8])>, Error> {
+    match Item::from_slice(input) {
```
Comment (djc) [inline src/pemfile.rs:81]: Nit: this looks like it's `Item::from_slice(input).map_err(Into::into)`?

File: src/pemfile.rs [hunk 5]
```diff
@@ -108,233 +94,20 @@ pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Er
...
-                        ErrorKind::InvalidData,
-                        format!(
-                            "illegal section start: {:?}",
-                            String::from_utf8_lossy(&line)
-                        ),
-                    ),
-
-                    Error::Base64Decode(err) => io::Error::new(ErrorKind::InvalidData, err),
-                });
+    match Item::from_buf(rd) {
```
Comment (djc) [inline src/pemfile.rs:97]: `map_err()` could work here, too?

File: src/pemfile.rs [hunk 6]
```diff
@@ -108,233 +94,20 @@ pub fn read_one_from_slice(mut input: &[u8]) -> Result<Option<(Item, &[u8])>, Er
...
+        Ok(out) => Ok(out),
+        Err(err) => match err {
+            parent::Error::Io(io) => Err(io),
+            other => {
+                let err: Error = other.into();
```
Comment (djc) [inline src/pemfile.rs:102]: Would suggest `Error::from(other)` instead.

**Review by djc — COMMENTED** [review]
This is cool!

File: src/pemfile.rs [hunk 7]
```diff
@@ -1,59 +1,17 @@
...
-    ///
-    /// Appears as "CERTIFICATE REQUEST" in PEM files.
-    Csr(CertificateSigningRequestDer<'static>),
-}
+use pki_types::pem as parent;
+pub use pki_types::pem::Item;
 
 /// Errors that may arise when parsing the contents of a PEM file
+///
+/// This differs from pki-types because it is `PartialEq`
```
Comment (cpu) [inline src/pemfile.rs:14]: Seems like we might as well use a rustdoc-link here and say  "This differs from [`pki-types::pem::Error`] ..". WDYT?

File: src/pemfile.rs [hunk 8]
```diff
@@ -72,29 +30,57 @@ pub enum Error {
...
+    fn from(error: parent::Error) -> Self {
+        match error {
+            parent::Error::MissingSectionEnd { end_marker } => {
+                Error::MissingSectionEnd { end_marker }
+            }
+            parent::Error::IllegalSectionStart { line } => Error::IllegalSectionStart { line },
+            parent::Error::Base64Decode(str) => Error::Base64Decode(str),
+
+            // this is a necessary bodge to funnel any new errors into our existing type
+            // (to which we can add no new variants)
```
Comment (cpu) [inline src/pemfile.rs:68]: maybe worth filing an issue with a "next release" label and citing it here?

File: src/pemfile.rs (hunk 1 above)
Comment (cpu) [reply src/pemfile.rs:9]: `use pki_types::pem` and then using `pem::` throughout seems reasonable to me. There's no` pem.rs` in this crate so I think it's fairly unambiguous.

File: src/lib.rs (hunk 2 above)
Comment (ctz) [reply src/lib.rs:107]: (fell away, it was due to the `Item` types `non_exhaustive`ness being activated by it being in a different crate)

File: src/pemfile.rs (hunk 8 above)
Comment (ctz) [reply src/pemfile.rs:156]: I was thinking we would probably not do a 3.x release for this crate?

Comment (djc) [reply src/pemfile.rs:156]: Would be good to avoid it, yeah.

File: src/pemfile.rs [hunk 9]
```diff
@@ -53,7 +51,57 @@ pub enum Item {
     Csr(CertificateSigningRequestDer<'static>),
 }
 
+impl Item {
+    fn from_kind(kind: pem::SectionKind, data: Vec<u8>) -> Option<Self> {
```
Comment (djc) [inline src/pemfile.rs:55]: This method should be on the bottom, since both `from_slice()` and `from_buf()` depend on it. Probably also sort `from_slice()` after `from_buf()`?

**Comment by ctz** [comment]
> I was hoping to rely on cargo-semver-checks here to check for API regressions, alas [obi1kenobi/cargo-semver-checks#609](https://github.com/obi1kenobi/cargo-semver-checks/issues/609) means this is not possible.

Now `Item` is back in this crate, this now works again and confirms (to the extent it can) there are no breaking changes.

**Review by djc — APPROVED** [review]
Okay, this seems to work out well?

I'm assuming you'll just squash all the commits here?

File: src/pemfile.rs (hunk 8 above)
Comment (cpu) [reply src/pemfile.rs:156]: WFM

**Review by cpu — APPROVED** [review]
Cool :+1:

**Review by djc — APPROVED** [review]
Nice!
